### PR TITLE
Fixed direct video links not rendering correctly.

### PR DESCRIPTION
**Pull request by brunoramoslu**
Fixed direct video links not rendering correctly.
Added webp and webm formats.

---

Direct video links[1][2] were not being rendered correctly and I were always showing a message:
`Your browser does not support the video tag.`

I did some changes to correctly create a `<div><video><source>` structure so that the video is rendered correctly.
Still need to update the tests in **basic.t** and clean up the `%for` in the **video.html.ep** template.
And the video is being rendered at 100%, so that needs some cleanup too (review class and width/height).

Could you please take a look already to see if you understand what I did?
Maybe provide some guidance/comments?

[1] https://www.learningcontainer.com/wp-content/uploads/2020/05/sample-mp4-file.mp4
[2] https://dl8.webmfiles.org/big-buck-bunny_trailer.webm

### CODE REVIEW

File: lib/LinkEmbedder/Link.pm [hunk 1]
```diff
@@ -382,9 +383,13 @@ __DATA__
...
-<video class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
-% for my $s (@{$l->{sources} || []}) {
-  <source src="<%= $s->{url} %>" type="<%= $s->{type} || '' %>">
-% }
-  <p>Your browser does not support the video tag.</p>
-</video>
+@@ video.html.ep
+<div class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
+  <video class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
+    <source src="<%= $l->{url} %>" type="<%= $l->{mimetype} || '' %>">
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link.pm:389]: I think it's better to use `$l->mimetype` instead of poking inside the hash.

File: lib/LinkEmbedder/Link.pm [hunk 2]
```diff
@@ -44,6 +44,7 @@ has ua               => undef;
 has url              => sub { Mojo::URL->new };                                            # Mojo::URL
 has version          => '1.0';
 has width            => sub { $_[0]->type =~ /^photo|video$/ ? 0 : undef };
+has mimetype         => '';
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link.pm:47]: Can you sort the attribute names alphabetically? Meaning moving this line up a couple of lines. The attribute also needs a "=head2 mimetype" in the POD section.

File: lib/LinkEmbedder/Link.pm [hunk 3]
```diff
@@ -382,9 +383,13 @@ __DATA__
...
 % }
 @@ video.html.ep
-<video class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
-% for my $s (@{$l->{sources} || []}) {
-  <source src="<%= $s->{url} %>" type="<%= $s->{type} || '' %>">
-% }
-  <p>Your browser does not support the video tag.</p>
-</video>
+@@ video.html.ep
+<div class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link.pm:387]: The div should not have height/width/preload/controls attributes.

File: lib/LinkEmbedder/Link.pm [hunk 4]
```diff
@@ -382,9 +383,13 @@ __DATA__
...
+  <video class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link.pm:388]: It will be a breaking change, but I think we should remove the classes here and just keep them on the div.

File: lib/LinkEmbedder/Link.pm [hunk 5]
```diff
@@ -382,9 +383,13 @@ __DATA__
...
+    <source src="<%= $l->{url} %>" type="<%= $l->{mimetype} || '' %>">
+    % for my $s (@{$l->{sources} || []}) {
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link.pm:390]: "sources" is not defined.

File: lib/LinkEmbedder/Link/Basic.pm [hunk 6]
```diff
@@ -14,6 +14,10 @@ sub learn_p {
 
   $self->type($type);
 
+  if ($type eq 'video' and $url =~ m/\.([^.]+)$/) {
+    $self->mimetype("video/$1");
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link/Basic.pm:18]: Maybe do `mimetype(lc "video/$1")` in case the extension is .WEBM, instead of .webm.

**Review by jhthorsen — CHANGES_REQUESTED** [review]
I think this looks great in general. Just need some tweaks.

Thanks for the contribution!

File: lib/LinkEmbedder/Link/Basic.pm (hunk 6 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link/Basic.pm:18]: Ok

File: lib/LinkEmbedder/Link.pm (hunk 5 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm:390]: I'm struggling to understand what this was supposed to do, I'm under the impression this was not working.
I'll take another look, and remove it if not needed.

File: lib/LinkEmbedder/Link.pm (hunk 3 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm:387]: Ok

File: lib/LinkEmbedder/Link.pm (hunk 2 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm:47]: Ok

File: lib/LinkEmbedder/Link.pm (hunk 1 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm:389]: Ok

File: lib/LinkEmbedder/Link.pm (hunk 4 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm:388]: Makes sense.
I'm wondering if we keep the height/width in the video tag or not...

Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm]: I removed the width/height from the video tag, and adding max-width and width to .le-video allows to control the size.

**Comment by brunoramoslu** [comment]
I did a round of changes and fixed the `basic.t` tests.
I'm checking it it was me who broke the other test... I'm having failures when using TEST_ONLINE=1 on some others tests.

File: t/basic.t [hunk 7]
```diff
@@ -55,11 +55,29 @@ LinkEmbedder->new->test_ok(
...
+        provider_url     => 'https://www.learningcontainer.com/',
+        thumbnail_height => undef,
+        thumbnail_url    => undef,
+        thumbnail_width  => undef,
+        title            => 'sample-mp4-file.mp4',
+        type             => 'video',
+        url              => 'https://www.learningcontainer.com/wp-content/uploads/2020/05/sample-mp4-file.mp4',
+        version          => '1.0'
+    }
+);
```
Comment (jhthorsen) [inline t/basic.t:79]: Awesome!

File: lib/LinkEmbedder/Link.pm [hunk 8]
```diff
@@ -382,9 +387,9 @@ __DATA__
...
-<video class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>" height="640" width="480" preload="metadata" controls>
-% for my $s (@{$l->{sources} || []}) {
-  <source src="<%= $s->{url} %>" type="<%= $s->{type} || '' %>">
-% }
-  <p>Your browser does not support the video tag.</p>
-</video>
+<div class="le-<%= $l->type %> le-provider-<%= lc $l->provider_name %>">
+  <video preload="metadata" controls>
+    <source src="<%= $l->url %>" type="<%= $l->mimetype || '' %>">
+    <p>Your browser does not support the video tag.</p>
```
Comment (jhthorsen) [inline lib/LinkEmbedder/Link.pm:393]: This is not really related to your PR, but I was wondering if this message should be changed into something like this:

```
<p>Your browser is unable to play the <%= $l->mimetype || 'video' %> element.</p>
```

**Review by jhthorsen — CHANGES_REQUESTED** [review]

File: main.pl [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+# pages. This is done automatically by $embedder->serve()
+$embedder->ua->transactor->name("Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36");
+
+$embedder->get_p("https://www.learningcontainer.com/wp-content/uploads/2020/05/sample-mp4-file.mp4")->then(sub {
+#$embedder->get_p("https://dl8.webmfiles.org/big-buck-bunny_trailer.webm")->then(sub {
+#$embedder->get_p("https://www.learningcontainer.com/mp4-sample-video-files-download/")->then(sub {
+    my $link = shift;
+    print $link->html;
+})->wait;
+
```
[the remark this answers is not in the document]
Comment (brunoramoslu) [reply main.pl]: Sorry about that.
Done.

File: lib/LinkEmbedder/Link.pm (hunk 5 above)
Comment (jhthorsen) [reply lib/LinkEmbedder/Link.pm]: Sorry about this one. I thought this was something you added, but I got tricked by the indentation.

**Review by jhthorsen — APPROVED** [review]
I'll rebase and merge this once you have pressed the "Ready for review" button 👍
Feel free to make more changes though, if you want to.

File: lib/LinkEmbedder/Link.pm (hunk 8 above)
Comment (brunoramoslu) [reply lib/LinkEmbedder/Link.pm:393]: Good idea.

**Comment by brunoramoslu** [comment]
I think I did most of what I was planning.
I'm glad to accommodate any other suggestions.
If there's nothing else you would like changed related to this, feel free to accept it.

I was wondering if there's something else needed to better integrate the changes in convos... but I do not know the code base well enough to give an informed opinion about that.

**Comment by jhthorsen** [comment]
> I was wondering if there's something else needed to better integrate the changes in convos... but I do not know the code base well enough to give an informed opinion about that.

Not sure what that would be, but please let me know if you have any specific ideas.
